File: src/runtime/boxes/BbBal.cs
// BbBal.cs — BAL: match a balanced-parentheses expression
// No direct C box (BAL is a primitive in SNOBOL4 spec).
// Semantics: match shortest string that is balanced with respect to
// parentheses '(' and ')'. No backtrack — matches exactly one balanced
// token at the current position.
//
// α: scan from cursor; count parens; succeed when depth returns to 0
//    and at least one char is consumed.
// β: ω

namespace Snobol4.Runtime.Boxes;

public sealed class BbBal : IByrdBox
{
    private int _matchedLen;

    public Spec Alpha(MatchState ms)
    {
        int pos   = ms.Cursor;
        int depth = 0;
        int len   = ms.Length;
        string s  = ms.Subject;

        if (pos >= len) return Spec.Fail;

        int start = pos;
        do
        {
            char c = s[pos];
            if      (c == '(') depth++;
            else if (c == ')')
            {
                if (depth == 0) return Spec.Fail;   // unmatched close
                depth--;
            }
            pos++;
        }
        while (pos < len && depth > 0);

        if (depth != 0) return Spec.Fail;            // unclosed open

        _matchedLen = pos - start;
        var result  = Spec.Of(start, _matchedLen);
        ms.Cursor   = pos;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor -= _matchedLen;
        return Spec.Fail;
    }
}

File: src/runtime/boxes/bal/bb_bal.java
package bb;

/**
 * bb_bal.java — BAL: match a balanced parenthesised string
 *
 * No bb_bal.c reference exists; semantics from SPITBOL/SNOBOL4 spec:
 *   Matches a string balanced with respect to '(' and ')'.
 *   Scans forward consuming chars; tracks depth. Succeeds at first
 *   position where depth returns to 0 after at least one char consumed.
 *   β: unconditional ω (BAL does not backtrack to shorter match).
 *
 *   BAL_α:  depth=0; len=0;
 *           scan: while (Δ+len < Ω):
 *             ch = Σ[Δ+len]
 *             if ch=='(' depth++
 *             else if ch==')' { if depth==0 → break; depth-- }
 *             len++;
 *             if depth==0 → BAL=spec(Σ+Δ,len); Δ+=len; goto BAL_γ;
 *           goto BAL_ω;
 *   BAL_β:  Δ -= len;  goto BAL_ω;
 */
public class bb_bal extends bb_box {
    private int len;   /* saved match length for β restore */

    public bb_bal(MatchState ms) { super(ms); }

    @Override public Spec α() {
        int depth = 0;
        len = 0;
        while (ms.delta + len < ms.omega) {
            char ch = ms.sigma.charAt(ms.delta + len);
            if      (ch == '(')              { depth++; len++; }
            else if (ch == ')' && depth > 0) { depth--; len++; }
            else if (ch == ')' && depth == 0){ break; }         /* unmatched ')' */
            else                             { len++; }
            if (depth == 0 && len > 0) {
                Spec r = new Spec(ms.delta, len);
                ms.delta += len;
                return r;                                                      // BAL_γ
            }
        }
        return null;                                                           // BAL_ω
    }

    @Override public Spec β() { ms.delta -= len; return null; }            // BAL_ω
}
